{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5db6c153-3a38-493b-a495-22fd2c0a6f93",
   "metadata": {},
   "outputs": [],
   "source": [
    "### blood methylation download wrangle \n",
    "\n",
    "# - genome build \n",
    "#     - GrCh38\n",
    "# - lab method:  \n",
    "#     - WG bisulfite seqeuncing --> turns cytosines into uracil but not methylated cytosines. then WGS to see where there are still Cs \n",
    "#     - read more : https://www.encodeproject.org/documents/964e2676-d0be-4b5d-aeec-f4f02310b221/@@download/attachment/WGBS%20pipeline%20overview.pdf\n",
    "# - file format & head: \n",
    "#     - OG: bigbed format (type of condesned binary) \n",
    "# - what the data represents: \n",
    "#     - Description of bedMethyl file: The bedMethyl file is a bed9+2 file containing the number of reads and the percent methylation. Each column represents the following: \n",
    "#         1. Reference chromosome or scaffold \n",
    "#         2. Start position in chromosome \n",
    "#         3. End position in chromosome \n",
    "#         4. Name of item \n",
    "#         5. Score from 0-1000. Capped number of reads \n",
    "#         6. Strandedness, plus (+), minus (-), or unknown (.) \n",
    "#         7. Start of where display should be thick (start codon) \n",
    "#         8. End of where display should be thick (stop codon) \n",
    "#         9. Color value (RGB) \n",
    "#         10. Coverage, or number of reads \n",
    "#         11. Percentage of reads that show methylation at this position in the genome \n",
    "\n",
    "\n",
    "#wget \n",
    "wget https://www.encodeproject.org/files/$2/@@download/$2.bigBed -P data/$1/track_data/methylation/ -o data/$1/track_data/methylation/wget_output_CHG.txt\n",
    "wget https://www.encodeproject.org/files/$3/@@download/$3.bigBed -P data/$1/track_data/methylation/ -o data/$1/track_data/methylation/wget_output_CHH.txt\n",
    "wget https://www.encodeproject.org/files/$4/@@download/$4.bigBed -P data/$1/track_data/methylation/ -o data/$1/track_data/methylation/wget_output_CpG.txt\n",
    "\n",
    "#convert to bed \n",
    "echo $1 \"converting to bed\" \n",
    "data/applications/UCSC_genomeBrowser_Blat/bigBedToBed data/$1/track_data/methylation/$2.bigBed data/$1/track_data/methylation/methylation_CHG_hg38.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/bigBedToBed data/$1/track_data/methylation/$3.bigBed data/$1/track_data/methylation/methylation_CHH_hg38.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/bigBedToBed data/$1/track_data/methylation/$4.bigBed data/$1/track_data/methylation/methylation_CpG_hg38.bed\n",
    "rm data/$1/track_data/methylation/$2.bigBed\n",
    "rm data/$1/track_data/methylation/$3.bigBed\n",
    "rm data/$1/track_data/methylation/$4.bigBed\n",
    "\n",
    "#prepare for liftover \n",
    "awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"|\"$5\"|\"$6\"|\"$7\"|\"$8\"|\"$9\"|\"$10\"|\"$11\"|\"$12\"|\"$13\"|\"$14}' data/$1/track_data/methylation/methylation_CHG_hg38.bed > data/$1/track_data/methylation/methylation_CHG_hg38_preppedLiftover.bed\n",
    "awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"|\"$5\"|\"$6\"|\"$7\"|\"$8\"|\"$9\"|\"$10\"|\"$11\"|\"$12\"|\"$13\"|\"$14}' data/$1/track_data/methylation/methylation_CHH_hg38.bed > data/$1/track_data/methylation/methylation_CHH_hg38_preppedLiftover.bed\n",
    "awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"|\"$5\"|\"$6\"|\"$7\"|\"$8\"|\"$9\"|\"$10\"|\"$11\"|\"$12\"|\"$13\"|\"$14}' data/$1/track_data/methylation/methylation_CpG_hg38.bed > data/$1/track_data/methylation/methylation_CpG_hg38_preppedLiftover.bed\n",
    "rm data/$1/track_data/methylation/methylation_CHG_hg38.bed\n",
    "rm data/$1/track_data/methylation/methylation_CHH_hg38.bed\n",
    "rm data/$1/track_data/methylation/methylation_CpG_hg38.bed\n",
    "\n",
    "# #liftvoer hg38-->19 \n",
    "echo $1 \"lifting over hg38-->hg19\"\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CHG_hg38_preppedLiftover.bed analysis/modules/download_encode/hg38ToHg19.over.chain.gz data/$1/track_data/methylation/methylation_CHG_hg19.bed  data/$1/track_data/methylation/methylation_CHG_unliftgedhg38.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CHH_hg38_preppedLiftover.bed analysis/modules/download_encode/hg38ToHg19.over.chain.gz data/$1/track_data/methylation/methylation_CHH_hg19.bed  data/$1/track_data/methylation/methylation_CHH_unliftgedhg38.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CpG_hg38_preppedLiftover.bed analysis/modules/download_encode/hg38ToHg19.over.chain.gz data/$1/track_data/methylation/methylation_CpG_hg19.bed  data/$1/track_data/methylation/methylation_CpG_unliftgedhg38.bed\n",
    "rm data/$1/track_data/methylation/methylation_CHG_hg38_preppedLiftover.bed \n",
    "rm data/$1/track_data/methylation/methylation_CHH_hg38_preppedLiftover.bed\n",
    "rm data/$1/track_data/methylation/methylation_CpG_hg38_preppedLiftover.bed\n",
    "\n",
    "# #liftvoer hg19-->18 \n",
    "echo $1 \"lifting over hg19-->hg18\"\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CHG_hg19.bed analysis/modules/download_encode/hg19ToHg18.over.chain.gz data/$1/track_data/methylation/methylation_CHG_hg18.bed  data/$1/track_data/methylation/methylation_CHG_unliftgedhg19.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CHH_hg19.bed analysis/modules/download_encode/hg19ToHg18.over.chain.gz data/$1/track_data/methylation/methylation_CHH_hg18.bed  data/$1/track_data/methylation/methylation_CHH_unliftgedhg19.bed\n",
    "data/applications/UCSC_genomeBrowser_Blat/liftOver data/$1/track_data/methylation/methylation_CpG_hg19.bed analysis/modules/download_encode/hg19ToHg18.over.chain.gz data/$1/track_data/methylation/methylation_CpG_hg18.bed  data/$1/track_data/methylation/methylation_CpG_unliftgedhg19.bed\n",
    "rm data/$1/track_data/methylation/methylation_CHG_hg19.bed \n",
    "rm data/$1/track_data/methylation/methylation_CHH_hg19.bed\n",
    "rm data/$1/track_data/methylation/methylation_CpG_hg19.bed\n",
    "\n",
    "#bcombine and sort \n",
    "echo $1 \"combining \"\n",
    "cat data/$1/track_data/methylation/methylation_CHG_hg18.bed > data/$1/track_data/methylation/methylation.bed \n",
    "cat data/$1/track_data/methylation/methylation_CHH_hg18.bed >> data/$1/track_data/methylation/methylation.bed \n",
    "cat data/$1/track_data/methylation/methylation_CpG_hg18.bed >> data/$1/track_data/methylation/methylation.bed \n",
    "rm data/$1/track_data/methylation/methylation_CHG_hg18.bed \n",
    "rm data/$1/track_data/methylation/methylation_CHH_hg18.bed \n",
    "rm data/$1/track_data/methylation/methylation_CpG_hg18.bed \n",
    "\n",
    "echo $1 \"sorting\"\n",
    "sort -k1,1V -k2,2n -o data/$1/track_data/methylation/methylation.bed data/$1/track_data/methylation/methylation.bed \n",
    "#the sort resulted in \"chrom bloacks not continuous\" . apparently a probablem with sorting the chrom name. can add -V (-k1,1V) to enforce good sort \n",
    "\n",
    "#tabix \n",
    "echo $1 \"tabixing\"\n",
    "bgzip data/$1/track_data/methylation/methylation.bed\n",
    "tabix -s 1 -b 2 -e 3 data/$1/track_data/methylation/methylation.bed.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7936d77e-d941-4cf4-841c-208a43c77960",
   "metadata": {},
   "outputs": [],
   "source": [
    "site = [\"chr1\",1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "495dc81a-1de4-48e1-bebc-117b788aeae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t593688\t593689\t\"0\"|6|+|668445|668446|0,255,0|6|0|CCA|CCA|2\n",
      "chr1\t593689\t593690\t\"0\"|6|+|668446|668447|0,255,0|6|0|CAG|CAG|2\n",
      "chr1\t593694\t593695\t\"0\"|6|+|668451|668452|0,255,0|6|0|CCT|CCT|2\n",
      "chr1\t593695\t593696\t\"0\"|6|+|668452|668453|0,255,0|6|0|CTC|CTC|2\n",
      "chr1\t593697\t593698\t\"0\"|6|+|668454|668455|0,255,0|6|0|CTC|CTC|2\n",
      "chr1\t593699\t593700\t\"0\"|6|+|668456|668457|0,255,0|6|0|CAG|CAG|2\n",
      "chr1\t593707\t593708\t\"0\"|6|+|668464|668465|0,255,0|6|0|CAG|CAG|2\n",
      "chr1\t593720\t593721\t\"0\"|6|+|668477|668478|0,255,0|6|0|CCA|CCA|2\n",
      "chr1\t593721\t593722\t\"0\"|6|+|668478|668479|0,255,0|6|0|CAT|CAT|2\n",
      "chr1\t593724\t593725\t\"0\"|6|+|668481|668482|0,255,0|6|0|CAG|CAG|2\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat ../../../data/blood/track_data/methylation/methylation.bed.gz | head -n 10000 | tail -n 10 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c17ffd2-f1ae-4abb-adce-99faf6ee00c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "import pysam \n",
    "import gzip\n",
    "from numpy.random import choice\n",
    "import numpy as np "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "824339ee-4658-46ad-9a40-647c56f056f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename_tmp =\"../../../\"+\"data/global/sequence/chr1.fa.gz\"\n",
    "with gzip.open(filename_tmp, \"rt\") as handle:\n",
    "    chr1_seq = (AlignIO.read(handle,\"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7b2a556-8fb5-4534-9c3e-0ac3b192fdee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sites = choice(range(len(chr1_seq)), 100)\n",
    "sites = [['chr1', 1212518, 1],\n",
    " ['chr1', 1794355, 1],\n",
    " ['chr1', 1942590, 1],\n",
    " ['chr1', 2030862, 1],\n",
    " ['chr1', 2228180, 1],\n",
    " ['chr1', 2460795, 1],\n",
    " ['chr1', 2786869, 1],\n",
    " ['chr1', 3008881, 1],\n",
    " ['chr1', 3589975, 1],\n",
    " ['chr1', 4129241, 1]]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eeb4827d-b49f-4539-8cbe-007e004b69a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_surrounding_contexts = [0,100,10000]\n",
    "alignment = chr1_seq\n",
    "row=[]\n",
    "filename = \"../../../\"+'data/liver/track_data/methylation/methylation.bed.gz'\n",
    "for site in sites: \n",
    "    list_Cs_notInMethyl_data = []\n",
    "    for distance in list_of_surrounding_contexts: \n",
    "        num_reads_list = []\n",
    "        percent_methy_list = []\n",
    "        track_output = [record for record in pysam.Tabixfile(filename).fetch(site[0], site[1]-distance, site[1]+1+distance)]\n",
    "        if track_output: \n",
    "            if distance in [0,1]: \n",
    "                record_pres = False\n",
    "                next_door = False\n",
    "                for record in track_output: \n",
    "                    if int(record.split(\"\\t\")[1]) == site[1]: \n",
    "                        record_pres =True\n",
    "                        num_reads_at_site,percent_methylated = record.split(\"|\")[6:8]\n",
    "                        if int(num_reads_at_site)>=10: \n",
    "                            percent_methy_list.append(int(percent_methylated))\n",
    "                    elif int(record.split(\"\\t\")[1]) in [site[1]-1,site[1]+1]: next_door = True\n",
    "                if not record_pres and next_door==False: #the track_output is empty\n",
    "                    print(\"methylation error\",site, track_output, record_pres, next_door, record.split(\"\\t\")[1])\n",
    "                    assert 1==2\n",
    "            else: \n",
    "                for record in track_output:\n",
    "                    num_reads_at_site,percent_methylated = record.split(\"|\")[6:8]\n",
    "                    if int(num_reads_at_site) >=10 :\n",
    "                        percent_methy_list.append(int(percent_methylated)) \n",
    "            if len(percent_methy_list) ==0:\n",
    "                row.append(\"noCoverage\")\n",
    "            else: \n",
    "                row.append(np.mean(percent_methy_list))\n",
    "        else: \n",
    "            if alignment[0][site[1]].upper() == \"C\": \n",
    "                list_Cs_notInMethyl_data.append(site)\n",
    "            row.append(\"no_data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6406f897-a592-4f2a-8e38-846d0c37dfce",
   "metadata": {},
   "source": [
    "# ok so the code that didnt work in the createDf notebook "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4a017367-fba6-4066-ba1d-0e2593993d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n",
      "[W::hts_idx_load3] The index file is older than the data file: ../../../data/liver/track_data/methylation/methylation.bed.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "list_of_surrounding_contexts = [0,100,10000]\n",
    "filename = \"../../../\"+'data/liver/track_data/methylation/methylation.bed.gz'\n",
    "row=[]\n",
    "for site in sites: \n",
    "#     alignment = fastas_dict[site[0]][0]\n",
    "    alignment = chr1_seq\n",
    "    list_Cs_notInMethyl_data = []\n",
    "    for distance in list_of_surrounding_contexts: \n",
    "        num_reads_list = []\n",
    "        percent_methy_list = []\n",
    "        track_output = [record for record in pysam.Tabixfile(filename).fetch(site[0], site[1]-distance, site[1]+1+distance)]\n",
    "        if track_output: \n",
    "            if distance in [0,1]: \n",
    "                record_pres = False\n",
    "                next_door = False\n",
    "                for record in track_output: \n",
    "                    if int(record.split(\"\\t\")[1]) == site[1]: \n",
    "                        record_pres =True\n",
    "                        num_reads_at_site,percent_methylated = record.split(\"|\")[6:8]\n",
    "                        if int(num_reads_at_site)>=10: \n",
    "                            percent_methy_list.append(int(percent_methylated))\n",
    "                    elif int(record.split(\"\\t\")[1]) in [site[1]-1,site[1]+1]: next_door = True\n",
    "                if not record_pres and next_door==False: #the track_output is empty\n",
    "                    print(\"methylation error\",site, track_output, record_pres, next_door, record.split(\"\\t\")[1])\n",
    "                    assert 1==2\n",
    "            else: \n",
    "                for record in track_output:\n",
    "                    num_reads_at_site,percent_methylated = record.split(\"|\")[6:8]\n",
    "                    if int(num_reads_at_site) >=10 :\n",
    "                        percent_methy_list.append(int(percent_methylated))\n",
    "            if len(percent_methy_list) == 0: \n",
    "                row.append(\"noCoverage\")\n",
    "            else: \n",
    "                row.append(np.mean(percent_methy_list))\n",
    "        else: \n",
    "            if alignment[0][site[1]].upper() == \"C\": \n",
    "                list_Cs_notInMethyl_data.append(site)\n",
    "            row.append(\"no_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e5ba058a-261a-4074-b6c2-8943360d916d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['noCoverage',\n",
       " 15.989795918367347,\n",
       " 12.655949853402083,\n",
       " 'no_data',\n",
       " 5.072916666666667,\n",
       " 6.550287356321839,\n",
       " 55.0,\n",
       " 2.8739495798319328,\n",
       " 8.995298100432574,\n",
       " 0.0,\n",
       " 18.62295081967213,\n",
       " 10.784448025785657,\n",
       " 0.0,\n",
       " 23.448979591836736,\n",
       " 11.199229252554222,\n",
       " 0.0,\n",
       " 4.192982456140351,\n",
       " 4.7658629993907216,\n",
       " 'noCoverage',\n",
       " 3.221052631578947,\n",
       " 8.062736273627364,\n",
       " 'no_data',\n",
       " 4.533898305084746,\n",
       " 7.58344531040043,\n",
       " 0.0,\n",
       " 4.072727272727272,\n",
       " 8.030449890270665,\n",
       " 0.0,\n",
       " 3.2903225806451615,\n",
       " 5.6776482513938165]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-mutability] *",
   "language": "python",
   "name": "conda-env-.conda-mutability-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
