{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0630879c-f3a7-4919-a6f8-c15833948f7e",
   "metadata": {},
   "source": [
    "# **prepping and wrangling data ready for model creation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "c3456f84-caed-4941-8415-4ae3c348ce2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(stringr)\n",
    "library(stringi)\n",
    "library(corrplot)\n",
    "library(rlang)\n",
    "library(ggplot2)\n",
    "\n",
    "\n",
    "tissue = \"germline\"\n",
    "model_name = \"model9\"\n",
    "equiv_toLowest=FALSE\n",
    "exclude_CpG=FALSE \n",
    "exclude_triplet=FALSE\n",
    "exclude_TCX_CCX = FALSE\n",
    "NA_omit=TRUE\n",
    "mappability_col_keep = FALSE\n",
    "\n",
    "# args = commandArgs(trailingOnly=TRUE)\n",
    "# tissue = args[1]\n",
    "# model_name = args[2]\n",
    "# if (\"_equiv_toLowest\" %in% args){ equiv_toLowest = TRUE\n",
    "# }else{equiv_toLowest = FALSE }\n",
    "# if (\"_noTriplets\" %in% args){exclude_triplet = TRUE\n",
    "# }else{ exclude_triplet = FALSE }\n",
    "# if (\"_noCpG\" %in% args){allTissueSpecTracks = TRUE\n",
    "# }else{allTissueSpecTracks = FALSE }\n",
    "# if (\"_noTCX_CCX\" %in% args){ exclude_TCX_CCX = TRUE\n",
    "# }else{ exclude_TCX_CCX = FALSE }\n",
    "# if (\"_allTissueSpecTracks\" %in% args){allTissueSpecTracks = TRUE\n",
    "# }else{allTissueSpecTracks = FALSE }\n",
    "\n",
    "tmp_file_path = \"../../../\"\n",
    "options(warn=1)\n",
    "\n",
    "model_desc_modify = \"\"\n",
    "\n",
    "#import the data \n",
    "input_filePath = paste(tmp_file_path,\"data/\",tissue,\"/dataframes/\",model_name,\"/predictorDf\",model_desc_modify,\".txt\",sep=\"\")\n",
    "all_data <- read.table(input_filePath, header = TRUE,sep=\"\\t\")\n",
    "\n",
    "#here we load in the mutation df so we can check the ref/triplet. also perfrom skin wrangling. \n",
    "if (tissue == \"germline\"){\n",
    "    mutations_df = read.table(paste(tmp_file_path,'data/germline/mutation_data/mutations_hg18_final.bed',sep=\"\"))\n",
    "    colnames(mutations_df) = c(\"Chromosome\",\"site\",\"site_fake_end\",\"ref\",\"alt\",\"Fathers_age_at_conception\",\"Mothers_age_at_conception\")\n",
    "}else if(tissue %in% c(\"blood\",\"liver\")){\n",
    "    mutations_df = read.table(paste(tmp_file_path,'data/',tissue,'/mutations/mutations.bed',sep=\"\"))\n",
    "    colnames(mutations_df )= c(\"Chromosome\",\"site\",\"site_fake_end\",\"ref\",\"alt\",\"ID\",\"VAF\",\"Gene name\", \"Region\", \"AA\", \"COSMIC\", \"Species\", \"Gender\", \"Age_in_years\",           \n",
    "                            \"Tissue/Cell type\",\"Single-cell_genomics_biotechnology_or_Method\",\"Control_sample_or_tissue\")\n",
    "}else if(tissue == \"skin\") {\n",
    "    mutations_df = read.table(paste(tmp_file_path,'data/skin/mutations/SomaMutDB/mutations_0based.bed',sep=\"\"))\n",
    "    colnames(mutations_df) = c(\"Chromosome\",\"site\",\"site_fake_end\",\"ref\",\"alt\",\"lab\",\"subtissue\")\n",
    "}else{ \n",
    "    print(paste(tissue,\" tissue specified not yet supported  !!~~~!!!~~~~!!\"))}\n",
    "\n",
    "#merging on the shared columns (chromosome and site) \n",
    "all_data <- merge(x = all_data, y = mutations_df, all.x = TRUE)\n",
    "\n",
    "#MAKING SURE THE reference base matches the middle of the triplet (all tissues) \n",
    "all_data$triplet_middle <- str_sub(all_data$triplet,2,2) \n",
    "all_data_muts = filter(all_data,all_data$mutation_status==1)\n",
    "all_data_muts = filter(all_data_muts,nchar(all_data_muts$ref)==1)#removing the rows where the \"ref\" column has more than 1 element (deletion) \n",
    "stopifnot(all_data_muts$triplet_middle == all_data_muts$ref)\n",
    "\n",
    "all_data <- all_data[,!(names(all_data) %in% c('site_fake_end','ref','alt',\"triplet_middle\",\"Fathers_age_at_conception\",\"Mothers_age_at_conception\",\n",
    "                                              \"ID\",\"VAF\",\"Gene name\", \"Region\", \"AA\", \"COSMIC\", \"Species\", \"Gender\", \"Age_in_years\",      \n",
    "                                            \"Tissue/Cell type\",\"Single-cell_genomics_biotechnology_or_Method\",\"Control_sample_or_tissue\"))]\n",
    "\n",
    "#skin txn wranfgling: two subtissue sto handle \n",
    "if (tissue == \"skin\"){\n",
    "    #making sure only non mutant sites have NA in the subtissue column \n",
    "    stopifnot(sum(all_data[!is.na(all_data$subtissue),]$mutation_status)==sum(all_data$mutation_status))\n",
    "    #extracting mut and non mut numbers  as well as fribro to kertino # within mutant sites \n",
    "    n_nonMuts = sum(all_data$mutation_status==0)\n",
    "    n_Muts = sum(all_data$mutation_status==1)\n",
    "    n_fibro = sum(all_data$subtissue==\"fibroblasts\",na.rm=TRUE)\n",
    "    n_keratino = sum(all_data$subtissue==\"keratinocytes\",na.rm=TRUE)\n",
    "    #getting the number keratino/fibro there should be in the non-mut sites to match the ratio within the muts \n",
    "    n_nonMuts_intoFibro = round(n_fibro*n_nonMuts/n_Muts)\n",
    "    n_nonMuts_intoKeratino = round(n_keratino*n_nonMuts/n_Muts)\n",
    "    stopifnot(n_nonMuts_intoFibro+ n_nonMuts_intoKeratino==n_nonMuts)\n",
    "    #extarcting the row indices of the non muts\n",
    "    nonMut_rowIndex <- c(1:nrow(all_data))[(is.na(all_data$subtissue))]\n",
    "    #selecting non mutant rows to be turned into firbroblasts \n",
    "    nonMuts_rowIndex_intoFibro <- sort(sample(nonMut_rowIndex,size = n_nonMuts_intoFibro, replace =FALSE))\n",
    "    # turning the row indexses sampled into fibroblasts. turning the remining NA into keratinoctyes \n",
    "    all_data$subtissue[nonMuts_rowIndex_intoFibro] <- \"fibroblasts\"\n",
    "    all_data$subtissue[is.na(all_data$subtissue)] <- \"keratinocytes\"\n",
    "    \n",
    "    #**ok so creating a column that is just transcription for skin**\n",
    "    #will be from the fibro or keratino column depending on the value in \"subtissue\" column\n",
    "    #https://stackoverflow.com/questions/48466652/combining-two-columns-conditionally-in-r\n",
    "    #https://www.sharpsightlabs.com/blog/case-when-r/\n",
    "    all_data <- all_data %>% mutate(Transcription.0=case_when(\n",
    "            subtissue==\"fibroblasts\" ~ Transcription_fibro.0,\n",
    "            TRUE ~ Transcription_kerat.0 ))\n",
    "    all_data <- all_data %>% mutate(Transcription.100=case_when(\n",
    "            subtissue==\"fibroblasts\" ~ Transcription_fibro.100,\n",
    "            TRUE ~ Transcription_kerat.100 ))\n",
    "    all_data <- all_data %>% mutate(Transcription.10000=case_when(\n",
    "            subtissue==\"fibroblasts\" ~ Transcription_fibro.10000,\n",
    "            TRUE ~ Transcription_kerat.10000 ))\n",
    "    #writing this df to the file before we remove columns (in case wqant o identify which lab came from etc )\n",
    "    write.csv(all_data,paste(tmp_file_path,\"data/skin/dataframes/\",model_name,\"/skin_labSubtissue_DF.csv\",sep=\"\"), row.names=FALSE)\n",
    "    #removing colnames so standard like the rest of the tissues \n",
    "    all_data <- all_data[,!(names(all_data) %in% c('Transcription_fibro.0','Transcription_fibro.100','Transcription_fibro.10000',\n",
    "                                                'Transcription_kerat.0','Transcription_kerat.100','Transcription_kerat.10000',\n",
    "                                               'lab','subtissue'))]\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if (equiv_toLowest==TRUE){\n",
    "    blood_nrow = nrow(read.table(paste(tmp_file_path,\"data/blood/dataframes/\",model_name,\"/predictorDf\",model_desc_modify,\".txt\",sep=\"\"),header = TRUE,sep=\"\\t\"))\n",
    "    all_data <-all_data[sample(nrow(all_data), blood_nrow), ]\n",
    "    model_desc_modify = paste(model_desc_modify,\"_equiv_toLowest\",sep=\"\")}else{\n",
    "    model_desc_modify = paste(model_desc_modify,\"_fullModel\",sep=\"\")}\n",
    "# if (\"_fullModel\" %in% args){model_desc_modify = paste(model_desc_modify,\"_fullModel\",sep=\"\")}\n",
    "\n",
    "\n",
    "if (exclude_CpG==TRUE){\n",
    "    all_data <-all_data[!str_detect(all_data$triplet,\"CG\"),]\n",
    "    model_desc_modify = paste(model_desc_modify,\"_noCpG\",sep=\"\")}\n",
    "if (exclude_TCX_CCX==TRUE){\n",
    "    model_desc_modify = paste(model_desc_modify,\"_noTCX_CCX\",sep=\"\")}\n",
    "\n",
    "# print(summary(as.factor(all_data$triplet)))\n",
    "if (exclude_triplet==TRUE){\n",
    "    all_data <- all_data[,!(names(all_data) %in% c('triplet'))]\n",
    "    model_desc_modify = paste(model_desc_modify,\"_noTriplets\",sep=\"\")}\n",
    "# print(colnames(all_data))\n",
    "\n",
    "#setting up the error output file \n",
    "error_output_file = paste(tmp_file_path,\"data/\",tissue,\"/objects/\",model_name,\"/\",tissue,\"_create_model_text_output\",model_desc_modify,\".txt\",sep=\"\")\n",
    "cat(\"output for the create_model notebook\",file=error_output_file,sep=\"\\n\")\n",
    "\n",
    "#chnaging the 1-surround to 0 surround (--> due to a mess up, only blood )\n",
    "# if (tissue==\"blood\"){\n",
    "#     all_data <- all_data %>% \n",
    "#   rename(Apercent.0=Apercent.1,Gpercent.0=Gpercent.1,Cpercent.0=Cpercent.1,Tpercent.0=Tpercent.1,\n",
    "#          methylation_coverage.0=methylation_coverage.1,methylation_precent.0=methylation_precent.1,\n",
    "#          H3k27.0=H3k27.1,H3k27me3.0=H3k27me3.1,H3k4me1.0=H3k4me1.1,H3k4me3.0=H3k4me3.1,H3k36me3.0=H3k36me3.1,\n",
    "#          Transcription.0=Transcription.1,recombination.0=recombination.1,Repeats.0=Repeats.1,DNAse.0=DNAse.1,laminB1.0=laminB1.1)\n",
    "# }\n",
    "\n",
    "#remove the protein binding flag (if it exists in the version of the predictor data: not existant in later versions)\n",
    "all_data$annotation <- gsub(\"protein_binding\",\"not_transcribed\",all_data$annotation)#there are so few protein binding sites that we may as well omit \n",
    "\n",
    "#removing unamppable points \"\n",
    "mappable_mut_summary <- all_data %>% \n",
    "    group_by(mappability,mutation_status)%>%\n",
    "    summarise(count=n())\n",
    "unmapped_muts = (filter(mappable_mut_summary, (mappability == \"not\")&(mutation_status == 1))$count)\n",
    "unmapped_nonMuts = (filter(mappable_mut_summary, (mappability == \"not\")&(mutation_status == 0))$count)\n",
    "string_to_print = paste(tissue,\":\",unmapped_muts,\" mutations and\",unmapped_nonMuts,\"nonMut sites removed by mappability filter out of \",nrow(all_data),\"sites\",sep=\" \")\n",
    "cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "if (mappability_col_keep==FALSE){\n",
    "    all_data <- filter(all_data,mappability==\"mappable\")\n",
    "    all_data <- all_data[,!(names(all_data) %in% c(\"mappability\"))]\n",
    "}else{model_desc_modify=paste(model_desc_modify,\"_withMappableCol\",sep=\"\")}\n",
    "\n",
    "#handling gc content\"\n",
    "all_data <- all_data %>%\n",
    "    mutate(GC_content.0 = Gpercent.0+Cpercent.0) %>%\n",
    "    mutate(GC_content.100 = Gpercent.100+Cpercent.100) %>% \n",
    "    mutate(GC_content.10000 = Gpercent.10000+Cpercent.10000)\n",
    "all_data <- all_data[,!(names(all_data) %in% c('Apercent.0','Gpercent.0','Cpercent.0','Tpercent.0',\n",
    "                                       'Apercent.100','Gpercent.100','Cpercent.100','Tpercent.100',\n",
    "                                       'Apercent.10000','Gpercent.10000','Cpercent.10000','Tpercent.10000'))]\n",
    "\n",
    "#editing the triplets \n",
    "if (exclude_triplet==FALSE){\n",
    "    all_data$triplet <- toupper(all_data$triplet)\n",
    "    #filter for rows that dont have NNN as the triplet --> write the details to file \n",
    "    cat(paste(nrow(all_data[all_data$triplet == \"NNN\",]), \"rows removed due to N in triplet, \",sep=\" \"),file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "    all_data <- all_data[all_data$triplet != \"NNN\",]\n",
    "    cat(paste(nrow(all_data),\"rows left\",sep=\" \"),file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "}\n",
    "\n",
    "\n",
    "#create df to test coverage in the methylation download notebook \n",
    "bases = c(\"A\",\"T\",\"C\",\"G\")\n",
    "c_trips = c()\n",
    "for (b1 in bases){\n",
    "    for (b3 in bases){\n",
    "        c_trips <- append(c_trips,paste(b1,\"C\",b3,sep=\"\"))\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "all_data <- all_data[,!str_detect(colnames(all_data),\"methyl\")]\n",
    "# model_desc_modify = paste(model_desc_modify,\"_noMethyl\",sep=\"\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#how many mutations and sites \n",
    "#printing the mutation ratio to file \n",
    "string_to_print = paste(\"mut/normal ratio: \",nrow(all_data[all_data$mutation_status == \"1\",])/nrow(all_data),sep=\" \")\n",
    "cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "#printing the total mutations included to file \n",
    "string_to_print = paste(\"n muts = \", nrow(all_data[all_data$mutation_status == \"1\",]),sep=\" \")\n",
    "cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "#printing the total rows included to file \n",
    "string_to_print = paste(\"total nrow\", nrow(all_data),sep=\" \")\n",
    "cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "\n",
    "\n",
    "#converting 64-->32 triplets \n",
    "if (exclude_triplet==FALSE){\n",
    "    rc_removeAG <- function(dna){\n",
    "        middle_base = substr(dna, 2, 2)\n",
    "        if(middle_base %in% c(\"A\",\"G\")){\n",
    "            dna <- stri_reverse(chartr(\"acgtACGT\", \"tgcaTGCA\", dna))}\n",
    "        return(dna)\n",
    "    }#substring slicing https://www.johnmyleswhite.com/notebook/2009/02/25/text-processing-in-r/\n",
    "    all_data$triplet <- unlist(lapply(as.character(all_data$triplet),rc_removeAG)) #need unlistto turn the list into a vector \n",
    "}\n",
    "\n",
    "if (exclude_TCX_CCX==TRUE){\n",
    "    all_data <-all_data[!str_detect(all_data$triplet,\"[TC]C[ATCG]\"),]}\n",
    "\n",
    "#check for problems with levels / column tye etc. \n",
    "if (exclude_triplet==FALSE){\n",
    "    all_data$triplet <- as.character(all_data$triplet) #need to convert to char so the as.factor properly reducs to 64 levels after the removal of 'triplets\" longer than 3\n",
    "    string_to_print = paste(nrow(all_data[nchar(as.character(all_data$triplet))!=3,]),\"rows removed due to triplet larger than 3 in length \",sep=\" \")\n",
    "    cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "    all_data <- all_data[nchar(all_data$triplet)==3,] #make sure only including triplets \n",
    "    string_to_print=paste(nrow(all_data),\" rows left\", sept = \" \")\n",
    "    cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "}\n",
    "\n",
    "#editing the annotation colun --> ignored into non_transcribed \n",
    "all_data$annotation <- gsub(\"ignored\",\"not_transcribed\",all_data$annotation)\n",
    "all_data$annotation <- gsub(\"protein_binding\",\"not_transcribed\",all_data$annotation)#there are so few protein binding sites that we may as well omit \n",
    "#rerplacing with \"not_transcribed\" is fine as they are the last level before non_transcribed in the annotation module \n",
    "\n",
    "#factorizing the columns \n",
    "all_data$mutation_status <- as.factor(all_data$mutation_status)\n",
    "if (exclude_triplet==FALSE){all_data$triplet <- as.factor(as.character(all_data$triplet))}\n",
    "all_data$Chromosome <- as.factor(all_data$Chromosome)\n",
    "all_data$annotation <- as.factor(all_data$annotation)\n",
    "all_data$CpGisland <- as.factor(all_data$CpGisland)\n",
    "\n",
    "\n",
    "\n",
    "#checking that the factor variables are correct (mutation status, triplets, chroms) --> will raise error if not true \n",
    "stopifnot(length(levels(all_data$mutation_status)) == 2)\n",
    "if (exclude_triplet==FALSE){\n",
    "    if(exclude_CpG==TRUE&exclude_TCX_CCX==FALSE){\n",
    "        stopifnot(length(levels(all_data$triplet))==28)}\n",
    "    if(exclude_CpG==FALSE&exclude_TCX_CCX==TRUE){\n",
    "        stopifnot(length(levels(all_data$triplet))==24)}\n",
    "    if(exclude_CpG==TRUE&exclude_TCX_CCX==TRUE){\n",
    "        stopifnot(length(levels(all_data$triplet))==22)}\n",
    "    if(exclude_CpG==FALSE&exclude_TCX_CCX==FALSE){\n",
    "        stopifnot(length(levels(all_data$triplet))==32)}}\n",
    "stopifnot(length(levels(all_data$Chromosome))==22)\n",
    "stopifnot(length(levels(all_data$CpGisland))==3)\n",
    "\n",
    "\n",
    "#if germline, then edit out the female and make the male the default ( i used to include both female and male) \n",
    "if (tissue == \"germline\"){\n",
    "    all_data <- select(all_data,-matches(\"female\")) \n",
    "    colnames(all_data)<- str_replace_all(colnames(all_data), \"_male\", \"\")\n",
    "}\n",
    "\n",
    "#na omit and checking how many rows were removed \n",
    "string_to_print = paste(nrow(all_data)- nrow(na.omit(all_data)), \"lost to NA values, \", nrow(na.omit(all_data)),\" rows remain after\",sep=\" \")\n",
    "cat(string_to_print,file=error_output_file,sep=\"\\n\",append=TRUE)\n",
    "if (NA_omit ==TRUE){\n",
    "    all_data <- na.omit(all_data)\n",
    "    }\n",
    "\n",
    "#writing frequencies and counts for the categories of factor predictors \n",
    "#get he factor names \n",
    "factor_names <- names(Filter(is.factor,all_data))\n",
    "factor_names <-factor_names[ !factor_names == 'mutation_status']\n",
    "#for loop to print for each factor \n",
    "predictor_freqs_count_filePath <- paste(tmp_file_path,\"data/\",tissue,\"/objects/\",model_name,\"/\",tissue,\"_create_model_predFreqs\",model_desc_modify,\".csv\",sep=\"\")\n",
    "cat(\"predictor,category,frequency,n_sites\",file=predictor_freqs_count_filePath,sep=\"\\n\")\n",
    "for (predictor_name in factor_names){\n",
    "#     string_to_print<-(paste(predictor_name,\" mutation distributions\",sep=\" \"))\n",
    "#     cat(string_to_print,file=predictor_freqs_count_filePath,sep=\"\\n\",append=TRUE)\n",
    "    test_df<-as.data.frame(table(all_data[,\"mutation_status\"], all_data[,predictor_name]))#https://stackoverflow.com/questions/14596420/how-to-get-value-by-column-name-in-r\n",
    "    colnames(test_df) <- c(\"mut_stat\",predictor_name,\"Freq\")\n",
    "    for (cur_annot in levels(all_data[,predictor_name])){\n",
    "        n_sites = sum(test_df[test_df[predictor_name]==cur_annot,]$Freq)\n",
    "        n_muts <- as.integer(test_df[test_df[predictor_name]==cur_annot&test_df[\"mut_stat\"]==\"1\"][3])#the third colun is the one with the count \n",
    "        string_to_print <-((paste(predictor_name,cur_annot,n_muts/n_sites,n_sites,sep=\",\")))\n",
    "        cat(string_to_print,file=predictor_freqs_count_filePath,sep=\"\\n\",append=TRUE)}}\n",
    "\n",
    "\n",
    "#print(\"about to do the distributions\")\n",
    "\n",
    "#distributions ~~~~~~~~~~~~~~~~~~~~~\n",
    "#getting only numeric \n",
    "num_cols <- colnames(all_data[,unlist(lapply(all_data, is.numeric), use.names = FALSE)])#https://stackoverflow.com/questions/5863097/selecting-only-numeric-columns-from-a-data-frame\n",
    "\n",
    "pdf(paste(paste(tmp_file_path,\"analysis/\",tissue,\"/plots/\",model_name,\"/\",tissue,\"_predictor_distributions\",model_desc_modify,\".pdf\", sep=\"\")))\n",
    "for (cur_pred in num_cols){\n",
    "    cur_data = all_data[,cur_pred]\n",
    "    title = paste(cur_pred,\"\\nsd:\",round(sd(cur_data),digits=2), \" mean:\", round(mean(cur_data),digits=2), \" median:\",round(median(cur_data),digits=2),\" min,max:\",round(min(cur_data),digits=5),\",\",max(round(cur_data,digits=2)))\n",
    "    hist(cur_data, main=title,breaks=20)\n",
    "}\n",
    "dev.off()\n",
    "pdf(paste(paste(tmp_file_path,\"analysis/\",tissue,\"/plots/\",model_name,\"/\",tissue,\"_predictor_distributions_remove0\",model_desc_modify,\".pdf\", sep=\"\")))\n",
    "for (cur_pred in num_cols){\n",
    "    cur_data = filter(all_data,!!sym(cur_pred)>0)[,cur_pred]# get r to handle strign as value https://stackoverflow.com/questions/48219732/pass-a-string-as-variable-name-in-dplyrfilter\n",
    "    title = paste(cur_pred,\"\\nsd:\",round(sd(cur_data),digits=2), \" mean:\", round(mean(cur_data),digits=2), \" median:\",round(median(cur_data),digits=2),\" min,max:\",round(min(cur_data),digits=5),\",\",max(round(cur_data,digits=2)))\n",
    "    hist(cur_data, main=title,breaks=20)\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "#making categorical columns into seperate numerical columns ready for standardizing \n",
    "all_data$dummy <- 1#need a dumym column that the model.matrix can remove (as it has to remove a column aparently)\n",
    "muts_col <- all_data$mutation_status#saving the column of mutations so i can add it back later\n",
    "if (tissue==\"global\"){\n",
    "    tissue_col <- all_data$tissue\n",
    "    non_num_preds = c(\"mutation_status\",\"tissue\")\n",
    "}else{non_num_preds = c(\"mutation_status\")}#saving the column of mutations (and tissue, if relevant) so i can add it back later\n",
    "\n",
    "preds_to_standardize<-!(names(all_data) %in% non_num_preds)\n",
    "all_data <-(data.frame((model.matrix(dummy~., all_data[, preds_to_standardize])[,-1] )))\n",
    "all_data$mutation_status <- muts_col\n",
    "if (tissue==\"global\"){\n",
    "    all_data$tissue <- tissue_col}\n",
    "\n",
    "#print(\"standardizing\")\n",
    "#STANDARDIZING~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "non_num_preds = c(\"mutation_status\")\n",
    "if (tissue==\"global\"){\n",
    "    non_num_preds = c(\"mutation_status\",\"tissue\")\n",
    "}\n",
    "preds_to_standardize<-!(names(all_data) %in% non_num_preds)#yes i have to do this again because i chnaged the order of the mutant column \n",
    "all_data[, preds_to_standardize]<- (as.data.frame(scale(all_data[, preds_to_standardize],center=TRUE,scale=TRUE)))\n",
    "\n",
    "\n",
    "#CORRELATION ANALYSIS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#print(\"creating and saving the corplot \")\n",
    "# num_data <- na.omit(subset(all_data, select=-c(Chromosome,triplet,annotation,CpGisland))) --> dont need this anymore as correlating ther categoricals too \n",
    "num_data <-all_data[,!(names(all_data) %in% non_num_preds)]\n",
    "num_data$mutation_status <- as.numeric(all_data$mutation_status)\n",
    "num_cor <- cor(num_data)\n",
    "filename = paste(tmp_file_path,\"analysis/\",tissue,\"/plots/\",model_name,\"/\",tissue,\"_corPlot\",model_desc_modify,\".pdf\", sep=\"\")\n",
    "pdf(filename)\n",
    "corrplot(num_cor,method = \"color\",type=\"upper\",number.cex=0.5,tl.cex = 0.5,main=tissue)\n",
    "dev.off() \n",
    "\n",
    "\n",
    "#print(\"create the correlation df and pruning for new corplot\")\n",
    "cor_df <- as.data.frame(which(num_cor >=0.8,arr.ind=T))\n",
    "cor_df$col_name <- colnames(num_cor)[cor_df$col]  #insert tyhe col names \n",
    "testFunc <- function(a, b) round(num_cor[a,b],digits=3)#https://stackoverflow.com/questions/15059076/call-apply-like-function-on-each-row-of-dataframe-with-multiple-arguments-from-e\n",
    "cor_df$cor_val <- apply(cor_df[,c('row','col')], 1, function(y) testFunc(y['row'],y['col'])) #insert the value of the correlatnio \n",
    "#exclude the self corelations and convert row names into a col \n",
    "cor_df<- filter(as.data.frame(cor_df),row!=col)[,c(\"col_name\",\"cor_val\")]\n",
    "cor_df<- tibble::rownames_to_column(cor_df, \"row_name\")\n",
    "#reorder the var names into alphabetical order \n",
    "testFunc <- function(a,b) paste(str_sort(c(a,b)),sep=\"-\")[1]\n",
    "cor_df$first_name <- apply(cor_df,1,function(y) testFunc(y['row_name'],y['col_name']))\n",
    "testFunc <- function(a,b) paste(str_sort(c(a,b)),sep=\"-\")[2]\n",
    "cor_df$last_name <- apply(cor_df,1,function(y) testFunc(y['row_name'],y['col_name']))\n",
    "#subset cols\n",
    "cor_df <- cor_df[,c(\"first_name\",\"last_name\",\"cor_val\")]\n",
    "#doing soem strign replace to maker the distinct below applicable \n",
    "cor_df <- data.frame(lapply(cor_df, function(x) {gsub(\"1[.]1\", \"1\", x)}))\n",
    "cor_df <- data.frame(lapply(cor_df, function(x) {gsub(\"[.]2\", \"\", x)}))\n",
    "cor_df <- data.frame(lapply(cor_df, function(x) {gsub(\"0[.]1\", \"0\", x)}))\n",
    "#remove the duplicates and reorder for legibility \n",
    "cor_df <- distinct(cor_df)\n",
    "cor_df <- cor_df[order(cor_df$first_name),]\n",
    "#save to file \n",
    "filename = paste(tmp_file_path,\"data/\",tissue,\"/dataframes/\",model_name,\"/\",tissue,\"_fullModel_corTable_above0.8\",model_desc_modify,\".csv\", sep=\"\")\n",
    "write.csv(cor_df,filename, row.names = FALSE)\n",
    "                          \n",
    "#REMOVING CORRELATED VARIABLES \n",
    "correlated_vars = c('H3k27.1','H3k27.0','H3k27.100',\n",
    "                         'H3k4me1.0','H3k4me1.1','H3k4me1.100',\n",
    "                         'H3k4me3.0','H3k4me3.1','H3k4me3.100',\n",
    "                         'H3k27me3.0','H3k27me3.1','H3k27me3.100',\n",
    "                         'H3k36me3.0','H3k36me3.1','H3k36me3.100',\n",
    "                         \"Transcription.0\",\"Transcription.1\",\"Transcription.100\",\n",
    "                         \"recombination.0\",\"recombination.1\",\"recombination.100\",\n",
    "                         \"DNAse.0\",\"DNAse.1\",\"DNAse.10000\",\n",
    "                         \"GC_content.0\",\n",
    "                         \"Repeats.100\")\n",
    "all_data <- all_data[,!(names(all_data) %in% correlated_vars)]\n",
    "                          \n",
    "\n",
    "#VIF ANALYSIS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "x_data <- all_data[,!(names(all_data) %in% c(\"mutation_status\",\"mutation_status1\"))]\n",
    "pred_names <-1:length((colnames(x_data)))\n",
    "adj_rsqs <-1:length((colnames(x_data)))\n",
    "vifs <-1:length((colnames(x_data)))\n",
    "i=1\n",
    "for (predictor_name in (colnames(x_data))){\n",
    "    formula = paste(predictor_name,\"~.\",sep=\"\")#write the formula as a string to feed into the lm function \n",
    "    adj_rsq <- summary(lm(formula,data=x_data))$adj.r.squared#extract r sqared \n",
    "    vif <- 1/(1-adj_rsq)                   #calculate vif \n",
    "    pred_names[i] <- predictor_name        #save the predictor name to the vector \n",
    "    adj_rsqs[i] <- adj_rsq                 #save the adjisted r squared to the vector \n",
    "    vifs[i] <- vif                         #save the vif to a vecotr \n",
    "    i=i+1                                  #iterate the index for saving the current r^2/vif \n",
    "    }\n",
    "inter_cor_vif_df <- data.frame(pred_names,adj_rsqs,vifs)\n",
    "inter_cor_vif_df<- inter_cor_vif_df[order(-vifs),]\n",
    "\n",
    "#saving the model variables \n",
    "filename =paste(tmp_file_path,\"data/\",tissue,\"/dataframes/\",model_name,\"/\",tissue,\"_vif\",model_desc_modify,\".csv\", sep=\"\")\n",
    "write.csv(inter_cor_vif_df,filename) \n",
    "\n",
    "num_data <-all_data[,!(names(all_data) %in% non_num_preds)]\n",
    "num_data$mutation_status <- as.numeric(all_data$mutation_status)\n",
    "num_cor <- cor(num_data)\n",
    "filename = paste(tmp_file_path,\"analysis/\",tissue,\"/plots/\",model_name,\"/\",tissue,\"_corPlot_pruned\",model_desc_modify,\".pdf\", sep=\"\")\n",
    "pdf(filename)\n",
    "corrplot(num_cor,method = \"color\",type=\"upper\",number.cex=0.5,tl.cex = 0.5,main=tissue)\n",
    "dev.off() \n",
    "                          \n",
    "                          \n",
    "#saving the model variables \n",
    "filename =paste(tmp_file_path,\"data/\",tissue,\"/dataframes/\",model_name,\"/\",tissue,\"_all_data_readyForPrediction\",model_desc_modify,\".csv\", sep=\"\")\n",
    "write.csv(all_data,filename)    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:.conda-mutability] *",
   "language": "R",
   "name": "conda-env-.conda-mutability-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
